### PR TITLE
fix test timing issue and flapper caused by unnecessary pop/push

### PR DESCRIPTION
**Pull request by matthiashanel**
fixed timing issue in unit test.
added another unit test for lru.
fixed pop / push issue. this caused the lru list to be reordered, which tripped of test assumptions.

### CODE REVIEW

File: server/dirstore.go [hunk 1]
```diff
@@ -686,18 +686,15 @@ func (store *DirJWTStore) startExpiring(reCheck time.Duration, limit int64, evic
 			now := time.Now()
 			store.Lock()
 			if pq.Len() > 0 {
-				if it := heap.Pop(pq).(*jwtItem); it.expiration <= now.Unix() {
+				if it := pq.heap[0]; it.expiration <= now.Unix() {
```
Comment (derekcollison) [inline server/dirstore.go:689]: So you have second resolution only here, which I understand is fine based on usage. For tests with higher time based resolutions sometimes we do stuff like.

```go
now := time.Now().Sub(900*time.Millisecond)
```

This allows the test to run in ~100ms versus over a second. Just food for thought.

File: server/dirstore_test.go [hunk 2]
```diff
@@ -56,7 +57,7 @@ func require_NoError(t *testing.T, err error) {
 func require_Error(t *testing.T, err error) {
 	t.Helper()
 	if err == nil {
-		t.Errorf("require no error, but got: %v", err)
+		t.Errorf("require error, but got none")
```
Comment (derekcollison) [inline server/dirstore_test.go:60]: Why Errorf vs Fatalf? Different Go routine?

**Review by derekcollison — APPROVED** [review]
LGTM - a few comments.

File: server/dirstore_test.go (hunk 2 above)
Comment (kozlovic) [reply server/dirstore_test.go:60]: Agreed. I usually don't use anything but Fatal() because there is no point in continuing the test.

**Review by kozlovic — APPROVED** [review]
LGTM

File: server/dirstore.go (hunk 1 above)
Comment (matthiashanel) [reply server/dirstore.go:689]: unfortunately I am stuck with second resolution due to jwt

Comment (derekcollison) [reply server/dirstore.go:689]: Understood, but you can still use the trick above to speed up the test.

File: server/dirstore_test.go (hunk 2 above)
Comment (matthiashanel) [reply server/dirstore_test.go:60]: not sure, this code was moved from account server and originally dependent on the require library.
probably at the time I saw t.Errorf more often.

Fixed for the entire file

File: server/dirstore_test.go [hunk 3]
```diff
@@ -436,28 +437,27 @@ func TestExpiration(t *testing.T) {
...
-	for i := 1; i <= 3; i++ {
-		account(i)
-		nh := dirStore.Hash()
-		require_NotEqual(t, h, nh)
-		h = nh
-	}
-	time.Sleep(500 * time.Millisecond)
-	for i := 3; i > 0; i-- {
+	failAt := time.Now().Add(4 * time.Second)
+	for time.Now().Before(failAt) {
```
Comment (derekcollison) [inline server/dirstore_test.go:450]: So this makes this test run for at least 4 seconds?

File: server/dirstore_test.go [hunk 4]
```diff
@@ -436,28 +437,27 @@ func TestExpiration(t *testing.T) {
...
-		require_NotEqual(t, h, nh)
-		h = nh
+		if len(f) == 1 {
+			lh := dirStore.Hash()
+			require_Equal(t, string(hNoExp[:]), string(lh[:]))
+			return
+		}
 	}
-	assertStoreSize(t, dirStore, 0)
+	t.Errorf("Waited more than 4 seconds for the file with expiration 1 second to expire")
```
Comment (derekcollison) [inline server/dirstore_test.go:460]: t.Fatalf

File: server/dirstore_test.go (hunk 3 above)
Comment (kozlovic) [reply server/dirstore_test.go:450]: No, as you can see there is a return in the loop once the assertion is done. The 4sec is the max before reporting error

Comment (derekcollison) [reply server/dirstore_test.go:450]: So should only run 100ms from first sleep etc in optimum case yes?

File: server/dirstore.go (hunk 1 above)
Comment (kozlovic) [reply server/dirstore.go:689]: But this is actual code, not test code. Are you saying that we do that in server code in some places? I don't think so.

Comment (kozlovic) [reply server/dirstore.go:689]: in test code we sometimes change expiration time of a jwt, or some other things. we can sometimes access server internals and change values, but this is done in test code itself, not in server code.

Comment (matthiashanel) [reply server/dirstore.go:689]: are you suggesting to round down the current time? Is this done in other places of the code?

Comment (derekcollison) [reply server/dirstore.go:689]: No apologies, thought it was test code stuff.

File: server/dirstore_test.go (hunk 3 above)
Comment (matthiashanel) [reply server/dirstore_test.go:450]: about a second.
In part expiration in the jwt has to be bigger than issue at. and so we have to do rounding when issuing the jwt.

File: server/dirstore_test.go (hunk 4 above)
Comment (matthiashanel) [reply server/dirstore_test.go]: fixed, thanks

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**